### PR TITLE
Upgrade kustomize-in-kubectl to v4.1.2

### PR DESCRIPTION
**Pull request by monopole**
[v4.1.2]:https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv4.1.2
[v4.1.0]:https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv4.1.0

#### What type of PR is this?

/kind bug
/sig cli
/priority important-soon

#### What this PR does / why we need it:

This upgrades kustomize to depend on kustomize [v4.1.2].
It mitigates a concern that a user might download and use a
maliciously crafted remote configuration.

#### Special notes for your reviewer:

This is a `go.mod` change to upgrade the kustomize
dependency from v4.0.5 to [v4.1.2].

#### Does this PR introduce a user-facing change?

Yes with respect to kubectl _v1.21_, no with respect to earlier versions
of kubectl.  kustomize was last upgraded on 8 March 2021 (#98946).

If a user has a `kustomization.yaml` file with values defined for the
`helmChartInflationGenerator` field, and the user runs 
```
kubectl kustomize {target}
```
the user will be asked to _also_ specify the `--enable-helm` flag to
allow processing to continue.  Previously the `--enable-helm` flag
did not exist.  This is to increase awareness that `helm template...`
will be run as a subprocess to inflate a template.

Further, the `helmBin` subfield of `helmChartInflationGenerator` field
will be ignored. If the helm command is anything other than simply
`helm` (on the user's `PATH`), then the user must specify it via the
`--helm-command` flag.

The `helmChartInflationGenerator` was not read at all in kubectl
versions preceding v1.21.

```release-note
Upgrades functionality of `kubectl kustomize` as described at
https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv4.1.2
```

#### How was this PR made?

```
./hack/pin-dependency.sh sigs.k8s.io/kustomize/kyaml v0.10.17
./hack/pin-dependency.sh sigs.k8s.io/kustomize/cmd/config v0.9.10
./hack/pin-dependency.sh sigs.k8s.io/kustomize/api v0.8.8
./hack/pin-dependency.sh sigs.k8s.io/kustomize/kustomize/v4 v4.1.2

./hack/update-vendor.sh
./hack/update-internal-modules.sh 
./hack/lint-dependencies.sh 

rm ./_output/local/bin/linux/amd64/kubectl
make kubectl
./_output/local/bin/linux/amd64/kubectl help kustomize | grep enable-helm
      --enable-helm=false: Enable use of the Helm chart inflator generator.
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -532,10 +532,10 @@ replace (
...
 	rsc.io/pdf => rsc.io/pdf v0.1.1
 	sigs.k8s.io/apiserver-network-proxy/konnectivity-client => sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.15
-	sigs.k8s.io/kustomize/api => sigs.k8s.io/kustomize/api v0.8.5
-	sigs.k8s.io/kustomize/cmd/config => sigs.k8s.io/kustomize/cmd/config v0.9.7
-	sigs.k8s.io/kustomize/kustomize/v4 => sigs.k8s.io/kustomize/kustomize/v4 v4.0.5
-	sigs.k8s.io/kustomize/kyaml => sigs.k8s.io/kustomize/kyaml v0.10.15
+	sigs.k8s.io/kustomize/api => sigs.k8s.io/kustomize/api v0.8.7
+	sigs.k8s.io/kustomize/cmd/config => sigs.k8s.io/kustomize/cmd/config v0.9.10
+	sigs.k8s.io/kustomize/kustomize/v4 => sigs.k8s.io/kustomize/kustomize/v4 v4.1.1
+	sigs.k8s.io/kustomize/kyaml => sigs.k8s.io/kustomize/kyaml v0.10.17
```
Comment (liggitt) [inline go.mod:538]: There should be corresponding changes to the `require` directives. Instructions for updating dependencies are at the top of the file (`pin-dependency.sh` will update both require and replace directives)

Comment (liggitt) [reply go.mod:538]: oh, nevermind, it's not required directly from the top-level module

File: vendor/github.com/imdario/mergo/merge.go [hunk 2]
```diff
@@ -39,6 +41,8 @@ type Transformers interface {
 // short circuiting on recursive types.
 func deepMerge(dst, src reflect.Value, visited map[uintptr]*visit, depth int, config *Config) (err error) {
 	overwrite := config.Overwrite
+	overwriteWithEmptySrc := config.overwriteWithEmptyValue
+	config.overwriteWithEmptyValue = false
```
Comment (liggitt) [inline vendor/github.com/imdario/mergo/merge.go:45]: this change looks buggy... opened https://github.com/imdario/mergo/issues/187

Comment (monopole) [reply vendor/github.com/imdario/mergo/merge.go]: Rolled back to mergo v0.3.5, so no change in k8s mergo deps for now.

**Comment by pickfire** [comment]
Isn't this breaking changes such that `--enable-helm` is required now but previously it wasn't require? Shouldn't it require a new version to do these sort of breaking changes? Also, the error message was very confusing.
```
Error: accumulating resources: accumulation err='accumulating resources from '../../base': '/mnt/overlays/dev/base' must resolve to a file': recursed accumulation of path '/mnt/overlays/dev/base': accumulating resources: accumulation err='accumulating resources from '../../../resources/nats': '/mnt/resources/nats' must resolve to a file': recursed accumulation of path '/mnt/resources/nats': trouble configuring builtin HelmChartInflationGenerator with config: `
name: stan
releaseName: semut-stan
repo: https://nats-io.github.io/k8s/helm/charts/
valuesFile: values.yaml
version: 0.8.0
`: must specify --enable-helm
```

**Pull request merged**